apache/superset
# fix: Volatile datasource ordering in dashboard export

### SUMMARY


When using the export_dashboards command on the CLI the order of the datasources in the json is inconsistent so if you're trying to use git versioning theres a big diff from reordering. This PR implements sorting so this no longer happens.

Resolves: https://github.com/apache/superset/issues/19478

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


NA

### TESTING INSTRUCTIONS


I tried editing https://github.com/apache/superset/blob/f91f83d011b4c83576f5fff76d502f5616f2972d/tests/integration_tests/cli_tests.py#L76 like:

```python
    json_content = json.loads(contents)

    # check that datasources are ordered consistently
    assert json_content.get("datasources")[0].get("__SqlaTable__").get("database_id") == 1
```

to assert the order of the datasource ids but none of the existing dashboard fixtures have multiple datasources and I couldnt figure out how to add one with multiple datasources. So I didnt include the change to the test, if theres a way to stage a multi-datasource dashboard in the tests then I can gladly add a new test to do that!

To manually test prior to this PR:

1. Create multiple datasources in your Superset instance
2. Create a chart with each datasource
3. Add both charts to a dashboard
4. Export the dashboard using the CLI
5. Make a copy of the exported file
6. Export again and do a diff on the new vs copied file (you might need to export a few times before it changes because the ordering is random)

Following this PR there should never be a diff in the datasources if nothing changed.

### ADDITIONAL INFORMATION
